syntax = "proto3";
package protos.playbooks;

import "protos/base.proto";
import "google/protobuf/wrappers.proto";
import "google/protobuf/struct.proto";
import "protos/playbooks/playbook_commons.proto";
import "protos/playbooks/intelligence_layer/interpreter.proto";

import "protos/playbooks/decision_task_definitions/decision_task.proto";

import "protos/playbooks/source_task_definitions/cloudwatch_task.proto";
import "protos/playbooks/source_task_definitions/grafana_task.proto";
import "protos/playbooks/source_task_definitions/new_relic_task.proto";
import "protos/playbooks/source_task_definitions/datadog_task.proto";
import "protos/playbooks/source_task_definitions/eks_task.proto";
import "protos/playbooks/source_task_definitions/sql_database_task.proto";
import "protos/playbooks/source_task_definitions/api_call_task.proto";
import "protos/playbooks/source_task_definitions/bash_command_task.proto";
import "protos/playbooks/source_task_definitions/documentation_task.proto";


// Playbook Task Protos
message PlaybookTask {
  google.protobuf.UInt64Value id = 1;
  Source source = 2;
  google.protobuf.StringValue name = 3;
  google.protobuf.StringValue description = 4;
  google.protobuf.StringValue notes = 5;
  google.protobuf.StringValue created_by = 6;
  google.protobuf.Struct global_variable_set = 7;
  InterpreterType interpreter_type = 8;
  oneof task {
    PlaybookDocumentationTaskDefinition documentation_task = 101;
    PlaybookCloudwatchTask cloudwatch_task = 102;
    PlaybookGrafanaTask grafana_task = 103;
    PlaybookNewRelicTask new_relic_task = 104;
    PlaybookDatadogTask datadog_task = 105;
    SqlDataFetchTask clickhouse_data_fetch_task = 106;
    SqlDataFetchTask postgres_data_fetch_task = 107;
    PlaybookEksDataFetchTask eks_data_fetch_task = 108;
    SqlDataFetchTask sql_database_connection_data_fetch_task = 109;
    PlaybookApiCallTask api_call_task = 110;
    PlaybookBashCommandTask bash_command_task = 111;
  }
}

message PlaybookTaskExecutionLog {
  google.protobuf.UInt64Value id = 1;
  sfixed64 timestamp = 2;
  PlaybookTask task = 3;
  PlaybookTaskResult result = 4;
  Interpretation interpretation = 5;
}

// Playbook Step Protos
message PlaybookStep {
  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue name = 2;
  google.protobuf.StringValue description = 3;
  google.protobuf.StringValue notes = 4;
  repeated ExternalLink external_links = 5;
  InterpreterType interpreter_type = 6;
  repeated PlaybookTask tasks = 7;
}

message PlaybookStepExecutionLogV2 {
  google.protobuf.UInt64Value id = 1;
  sfixed64 timestamp = 2;
  google.protobuf.StringValue playbook_run_id = 3;
  PlaybookStep step = 4;
  repeated PlaybookTaskExecutionLog task_execution_logs = 5;
  Interpretation step_interpretation = 6;
}


message PlaybookDefinition {
  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue name = 2;
  google.protobuf.StringValue description = 3;
  google.protobuf.Struct global_variable_set = 4;
  google.protobuf.StringValue created_by = 5;
  google.protobuf.BoolValue is_active = 6;
  sfixed64 created_at = 7;
  sfixed64 last_run_at = 8;
  PlaybookExecutionStatusType status = 9;
  repeated PlaybookStep steps = 10;
}

message PlaybookExecutionV2 {
  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue playbook_run_id = 2;
  PlaybookDefinition playbook = 3;
  PlaybookExecutionStatusType status = 4;
  sfixed64 created_at = 5;
  sfixed64 started_at = 6;
  sfixed64 finished_at = 7;
  TimeRange time_range = 8;
  google.protobuf.StringValue created_by = 9;
  repeated PlaybookStepExecutionLogV2 step_execution_logs = 11;
}

// Crud Protos
message UpdatePlaybookOpV2 {
  enum Op {
    UNKNOWN = 0;
    UPDATE_PLAYBOOK_NAME = 1;
    UPDATE_PLAYBOOK_STATUS = 2;
    UPDATE_PLAYBOOK = 3;
  }

  message UpdatePlaybookName {
    google.protobuf.StringValue name = 1;
  }

  message UpdatePlaybookStatus {
    google.protobuf.BoolValue is_active = 1;
  }

  message UpdatePlaybook {
    PlaybookDefinition playbook = 1;
  }

  Op op = 1;
  oneof update {
    UpdatePlaybookName update_playbook_name = 2;
    UpdatePlaybookStatus update_playbook_status = 3;
    UpdatePlaybook update_playbook = 4;
  }
}